jsk_perception/src/draw_labeled_rect.cpp

```diff
@@ -0,0 +1,147 @@
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/o2r other materials provided
+ *     with the distribution.
+ *   * Neither the name of the JSK Lab nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+/*
+ * Author: Yuki Furuta <furushchev@jsk.imi.i.u-tokyo.ac.jp>
+ */
+
+#include <jsk_perception/draw_labeled_rect.h>
+
+namespace enc = sensor_msgs::image_encodings;
+
+namespace jsk_perception
+{
+  void DrawLabeledRect::onInit()
+  {
+    ConnectionBasedNodelet::onInit();
+
+    pnh_->param("approximate_sync", approximate_sync_, false);
+    pnh_->param("queue_size", queue_size_, 100);
+    pnh_->param("labels", labels_, std::vector<std::string>());
+
+    if (labels_.size() > 0) {
+      label_num_ = labels_.size();
+    } else {
+      label_num_ = 100;
+    }
+
+    pub_image_ = advertise<sensor_msgs::Image>(*pnh_, "output", 1);
+
+    onInitPostProcess();
+  }
+
+  void DrawLabeledRect::subscribe()
+  {
+    sub_image_.subscribe(*pnh_,   "input", 1);
+    sub_rects_.subscribe(*pnh_, "input/rect_array", 1);
+    if (approximate_sync_) {
+      async_rects_ = boost::make_shared<message_filters::Synchronizer<ApproximateSyncPolicyRects> >(queue_size_);
+      async_rects_->connectInput(sub_image_, sub_rects_);
+      async_rects_->registerCallback(boost::bind(&DrawLabeledRect::callbackRects, this, _1, _2));
+    } else {
+      sync_rects_ = boost::make_shared<message_filters::Synchronizer<ExactSyncPolicyRects> >(queue_size_);
+      sync_rects_->connectInput(sub_image_, sub_rects_);
+      sync_rects_->registerCallback(boost::bind(&DrawLabeledRect::callbackRects, this, _1, _2));
+    }
+  }
+
+  void DrawLabeledRect::unsubscribe()
+  {
+    sub_image_.unsubscribe();
+    sub_rects_.unsubscribe();
+  }
+
+  void DrawLabeledRect::getColorString(const jsk_recognition_msgs::Label &label, cv::Scalar& color)
+  {
+    static const float colors[6][3] = { {1,0,1}, {0,0,1},{0,1,1},{0,1,0},{1,1,0},{1,0,0} };
```

Comment:
Why not use:

- https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/jsk_recognition_utils/src/rgb_colors.cpp